### ISSUE TITLE
[BUG] MLFlow commands fails to execute concurrently

### ISSUE THREAD

**Issue by Tradunsky** [mlflow/mlflow#2201]
#1769 # System information
- **Have I written custom code (as opposed to using a stock example script provided in MLflow)**: Nope
- **OS Platform and Distribution (e.g., Linux Ubuntu 16.04)**: Jenkins agent, miniconda docker running on linuxkit
- **MLflow installed from (source or binary)**: binary
- **MLflow version (run ``mlflow --version``)**: 0.9.1 (but the same for 1.4.0)
- **Python version**: 2.7
- **npm version, if running the dev UI**:
- **Exact command to reproduce**: 
     mlflow run -e test ./test-model
     mlflow run -e build ./test-model

### Describe the problem
Running CI pipeline on jenkins, since tests and training takes long time I wanted to parallelize it. It fails to be executed concurrently, but passes on sequential call.

### Code to reproduce issue
     stage('Parallel') {
         failFast true
         parallel {
            stage('MLFlow test') {
                steps {
                       sh 'mlflow run -e test ./test-model'
                }
            }
            stage('MLFlow build') {
                steps {
                       sh 'mlflow run -e build ./test-model'                                                   
                }
            }
         }
     }

### Other info / logs
```+ mlflow run -e test ./test-model

2019/12/06 16:05:59 INFO mlflow.projects: === Creating conda environment mlflow-dee3cb9d8ec1d078ac23cda04ee09d2ead6fbbbd ===

Collecting package metadata (repodata.json): ...working... Traceback (most recent call last):

  File "/opt/conda/bin/mlflow", line 10, in <module>

    sys.exit(cli())

  File "/opt/conda/lib/python2.7/site-packages/click/core.py", line 764, in __call__

    return self.main(*args, **kwargs)

  File "/opt/conda/lib/python2.7/site-packages/click/core.py", line 717, in main

    rv = self.invoke(ctx)

  File "/opt/conda/lib/python2.7/site-packages/click/core.py", line 1137, in invoke

    return _process_result(sub_ctx.command.invoke(sub_ctx))

  File "/opt/conda/lib/python2.7/site-packages/click/core.py", line 956, in invoke

    return ctx.invoke(self.callback, **ctx.params)

  File "/opt/conda/lib/python2.7/site-packages/click/core.py", line 555, in invoke

    return callback(*args, **kwargs)

  File "/opt/conda/lib/python2.7/site-packages/mlflow/cli.py", line 139, in run

    run_id=run_id,

  File "/opt/conda/lib/python2.7/site-packages/mlflow/projects/__init__.py", line 230, in run

    storage_dir=storage_dir, block=block, run_id=run_id)

  File "/opt/conda/lib/python2.7/site-packages/mlflow/projects/__init__.py", line 132, in _run

    conda_env_name = _get_or_create_conda_env(project.conda_env_path)

  File "/opt/conda/lib/python2.7/site-packages/mlflow/projects/__init__.py", line 462, in _get_or_create_conda_env

    conda_env_path], stream_output=True)

  File "/opt/conda/lib/python2.7/site-packages/mlflow/utils/process.py", line 38, in exec_cmd

    raise ShellCommandException("Non-zero exitcode: %s" % (exit_code))

mlflow.utils.process.ShellCommandException: Non-zero exitcode: -9

script returned exit code 1

**Comment by Tradunsky** [comment]
It's not the first issue I've created being ignored, so I'm curious if mlflow community is alive?

We better not use this project if it has not active community thought.

**Comment by AveshCSingh** [comment]
Sorry for the late response on this, Slava. 

MLflow Projects don't specifically handle concurrent project execution.

Based on the logs you provided, it's not clear whether this is an issue with the code in your project or an MLflow issue. I'm closing this for now, but feel free to reopen with a fully reproducible example.

**Comment by Tradunsky** [comment]
Hi @AveshCSingh 
Sorry, I didn't get you - you want either me to type a new example that does execution concurrently on some language like python or you just saying mlflow run cli is not designed to to be executed concurrently?

In any cases, please keep it bound to the problem, since running mlflow on CI system is pretty basic use case.

**Comment by AveshCSingh** [comment]
Hi Slava,

Sorry for the confusion. The mlflow run cli should be safe to execute concurrently. I believe the problem here is actually that Conda is concurrently trying to create two environments with the same name. You can resolve this by creating the conda environment in the setup of your test scaffolding.

I've reproduced your issue:

```
$: mlflow run -e build ~/mlflow-example-project &  mlflow run -e test ~/mlflow-example-project
[1] 46014
/Users/avesh.singh/opt/anaconda2/envs/mlflow-dev-env/lib/python3.6/site-packages/pyspark/cloudpickle.py:47: DeprecationWarning: the imp module is deprecated in favour of importlib; see the module's documentation for alternative uses
  import imp
/Users/avesh.singh/opt/anaconda2/envs/mlflow-dev-env/lib/python3.6/site-packages/pyspark/cloudpickle.py:47: DeprecationWarning: the imp module is deprecated in favour of importlib; see the module's documentation for alternative uses
  import imp
2020/02/06 11:07:22 INFO mlflow.projects: === Creating conda environment mlflow-a334bf26be61bfb18b0f661cca4057c207a14948 ===
2020/02/06 11:07:22 INFO mlflow.projects: === Creating conda environment mlflow-a334bf26be61bfb18b0f661cca4057c207a14948 ===
Collecting package metadata (repodata.json): - Collecting package metadata (repodata.json): done
Solving environment: done
Solving environment: done
/ 

==> WARNING: A newer version of conda exists. <==
  current version: 4.7.12
  latest version: 4.8.2

Please update conda by running

    $ conda update -n base -c defaults conda


done
Preparing transaction: - 

==> WARNING: A newer version of conda exists. <==
  current version: 4.7.12
  latest version: 4.8.2

Please update conda by running

    $ conda update -n base -c defaults conda


\ Preparing transaction: done
Verifying transaction: done
Verifying transaction: done
Executing transaction: done
Executing transaction: done
/ #
# To activate this environment, use
#
#     $ conda activate mlflow-a334bf26be61bfb18b0f661cca4057c207a14948
#
# To deactivate an active environment, use
#
#     $ conda deactivate

\ 2020/02/06 11:07:38 INFO mlflow.projects: === Created directory /var/folders/tg/vkmyk0kj2615vnmh0tz_88wr0000gp/T/tmpmju6fhr0 for downloading remote URIs passed to arguments of type 'path' ===
2020/02/06 11:07:38 INFO mlflow.projects: === Running command 'source /Users/avesh.singh/opt/anaconda2/bin/../etc/profile.d/conda.sh && conda activate mlflow-a334bf26be61bfb18b0f661cca4057c207a14948 1>&2 && echo 'hello from build'' in run with ID '6c2cf470760847f48ea68824b0edc7d3' === 
| hello from build
2020/02/06 11:07:39 INFO mlflow.projects: === Run (ID '6c2cf470760847f48ea68824b0edc7d3') succeeded ===
done
ERROR conda.core.link:_execute(700): An error occurred while installing package 'defaults::numpy-base-1.18.1-py37h6575580_1'.
Rolling back transaction: done

CondaMultiError: CondaError: Cannot link a source that does not exist. /Users/avesh.singh/opt/anaconda2/envs/mlflow-a334bf26be61bfb18b0f661cca4057c207a14948/.condatmp/bc00a856-5e20-4b9c-a675-1a3c71cb5d99
Running `conda clean --packages` may resolve your problem.
()

Traceback (most recent call last):
  File "/Users/avesh.singh/opt/anaconda2/envs/mlflow-dev-env/bin/mlflow", line 11, in <module>
    load_entry_point('mlflow', 'console_scripts', 'mlflow')()
  File "/Users/avesh.singh/opt/anaconda2/envs/mlflow-dev-env/lib/python3.6/site-packages/click/core.py", line 764, in __call__
    return self.main(*args, **kwargs)
  File "/Users/avesh.singh/opt/anaconda2/envs/mlflow-dev-env/lib/python3.6/site-packages/click/core.py", line 717, in main
    rv = self.invoke(ctx)
  File "/Users/avesh.singh/opt/anaconda2/envs/mlflow-dev-env/lib/python3.6/site-packages/click/core.py", line 1137, in invoke
    return _process_result(sub_ctx.command.invoke(sub_ctx))
  File "/Users/avesh.singh/opt/anaconda2/envs/mlflow-dev-env/lib/python3.6/site-packages/click/core.py", line 956, in invoke
    return ctx.invoke(self.callback, **ctx.params)
  File "/Users/avesh.singh/opt/anaconda2/envs/mlflow-dev-env/lib/python3.6/site-packages/click/core.py", line 555, in invoke
    return callback(*args, **kwargs)
  File "/Users/avesh.singh/mlflow/mlflow/cli.py", line 134, in run
    run_id=run_id
  File "/Users/avesh.singh/mlflow/mlflow/projects/__init__.py", line 288, in run
    use_conda=use_conda, storage_dir=storage_dir, synchronous=synchronous, run_id=run_id)
  File "/Users/avesh.singh/mlflow/mlflow/projects/__init__.py", line 162, in _run
    conda_env_name = _get_or_create_conda_env(project.conda_env_path)
  File "/Users/avesh.singh/mlflow/mlflow/projects/__init__.py", line 453, in _get_or_create_conda_env
    conda_env_path], stream_output=True)
  File "/Users/avesh.singh/mlflow/mlflow/utils/process.py", line 38, in exec_cmd
    raise ShellCommandException("Non-zero exitcode: %s" % (exit_code))
mlflow.utils.process.ShellCommandException: Non-zero exitcode: 1
[1]+  Done                    mlflow run -e build ~/mlflow-example-project
```

The issue does not occur if the Conda environment has already been created:
```
(base) $: conda info --envs
# conda environments:
#
base                  *  /Users/avesh.singh/opt/anaconda2
mlflow-a334bf26be61bfb18b0f661cca4057c207a14948     /Users/avesh.singh/opt/anaconda2/envs/mlflow-a334bf26be61bfb18b0f661cca4057c207a14948
mlflow-dev-env           /Users/avesh.singh/opt/anaconda2/envs/mlflow-dev-env
```
```
$: mlflow run -e build ~/mlflow-example-project &&  mlflow run -e test ~/mlflow-example-project
2020/02/06 11:12:26 INFO mlflow.projects: === Creating conda environment mlflow-a334bf26be61bfb18b0f661cca4057c207a14948 ===
Collecting package metadata (repodata.json): done
Solving environment: done


==> WARNING: A newer version of conda exists. <==
  current version: 4.7.12
  latest version: 4.8.2

Please update conda by running

    $ conda update -n base -c defaults conda



Downloading and Extracting Packages
ca-certificates-2020 | 125 KB    | ######################################################################################################################################################################################################### | 100% 
mkl-service-2.3.0    | 201 KB    | ######################################################################################################################################################################################################### | 100% 
setuptools-45.1.0    | 511 KB    | ######################################################################################################################################################################################################### | 100% 
six-1.14.0           | 26 KB     | ######################################################################################################################################################################################################### | 100% 
python-3.7.0         | 13.2 MB   | ######################################################################################################################################################################################################### | 100% 
openssl-1.0.2u       | 2.0 MB    | ######################################################################################################################################################################################################### | 100% 
sqlite-3.31.1        | 1.2 MB    | ######################################################################################################################################################################################################### | 100% 
blas-1.0             | 6 KB      | ######################################################################################################################################################################################################### | 100% 
numpy-1.18.1         | 5 KB      | ######################################################################################################################################################################################################### | 100% 
Preparing transaction: done
Verifying transaction: done
Executing transaction: done
#
# To activate this environment, use
#
#     $ conda activate mlflow-a334bf26be61bfb18b0f661cca4057c207a14948
#
# To deactivate an active environment, use
#
#     $ conda deactivate

2020/02/06 11:12:53 INFO mlflow.projects: === Created directory /var/folders/tg/vkmyk0kj2615vnmh0tz_88wr0000gp/T/tmpWT9y6B for downloading remote URIs passed to arguments of type 'path' ===
2020/02/06 11:12:53 INFO mlflow.projects: === Running command 'source /Users/avesh.singh/opt/anaconda2/bin/../etc/profile.d/conda.sh && conda activate mlflow-a334bf26be61bfb18b0f661cca4057c207a14948 1>&2 && echo 'hello from build'' in run with ID '063d85382191412ab233d43e132edd58' === 
hello from build
2020/02/06 11:12:53 INFO mlflow.projects: === Run (ID '063d85382191412ab233d43e132edd58') succeeded ===
2020/02/06 11:12:57 INFO mlflow.projects: === Created directory /var/folders/tg/vkmyk0kj2615vnmh0tz_88wr0000gp/T/tmpKBqgu7 for downloading remote URIs passed to arguments of type 'path' ===
2020/02/06 11:12:57 INFO mlflow.projects: === Running command 'source /Users/avesh.singh/opt/anaconda2/bin/../etc/profile.d/conda.sh && conda activate mlflow-a334bf26be61bfb18b0f661cca4057c207a14948 1>&2 && echo 'hello from test'' in run with ID 'adc2d1efadac47689a107a10ff1fa517' === 
hello from test
2020/02/06 11:12:58 INFO mlflow.projects: === Run (ID 'adc2d1efadac47689a107a10ff1fa517') succeeded ===
```

**Comment by Tradunsky** [comment]
Hi @AveshCSingh 
Thank you for taking time to reproduce the issue.

I think that you're right that mlflow tries to create conda env with the same name, I looked into sources and it generates name by hashing conda env content.

Since mlflow defined different types of envs and makes decision on conda env re-usage, it is better to expose mlflow (CLI and libs) method to just create env. Other commands already have an option to ommit conda creation, so users might want to split conda creation and mlflow run commands.

btw, Seems like in the second example you executed it sequentially.

**Comment by Tradunsky** [comment]
Also, having re-using conda environment by name is not ideal, since conda env can fail to be fully created, like the conda env created, but one of dependencies failed to be installed, this leads to another defect - mlflow run on invalid conda environment.

I suggest to remove conda environment if it failed to be created, this way mlflow could avoid re-suing invalid env or sophisticated conda env validations.

WDYT?

**Comment by AveshCSingh** [comment]
Ah my mistake, I ran the commands in concurrently but copy-pasted the incorrect output. It should be:

```
(base) $: mlflow run -e build ~/mlflow-example-project & mlflow run -e test ~/mlflow-example-project
[1] 72371
2020/02/06 16:53:52 INFO mlflow.projects: === Created directory /var/folders/tg/vkmyk0kj2615vnmh0tz_88wr0000gp/T/tmpKfqcZ0 for downloading remote URIs passed to arguments of type 'path' ===
2020/02/06 16:53:52 INFO mlflow.projects: === Created directory /var/folders/tg/vkmyk0kj2615vnmh0tz_88wr0000gp/T/tmplfAV3k for downloading remote URIs passed to arguments of type 'path' ===
2020/02/06 16:53:52 INFO mlflow.projects: === Running command 'source /Users/avesh.singh/opt/anaconda2/bin/../etc/profile.d/conda.sh && conda activate mlflow-a334bf26be61bfb18b0f661cca4057c207a14948 1>&2 && echo 'hello from test'' in run with ID '18bc7e80c6b540dda738097a100ba607' === 
2020/02/06 16:53:52 INFO mlflow.projects: === Running command 'source /Users/avesh.singh/opt/anaconda2/bin/../etc/profile.d/conda.sh && conda activate mlflow-a334bf26be61bfb18b0f661cca4057c207a14948 1>&2 && echo 'hello from build'' in run with ID '8417c45a37304573a506078e39d47f91' === 
hello from build
hello from test
2020/02/06 16:53:52 INFO mlflow.projects: === Run (ID '8417c45a37304573a506078e39d47f91') succeeded ===
2020/02/06 16:53:52 INFO mlflow.projects: === Run (ID '18bc7e80c6b540dda738097a100ba607') succeeded ===
[1]+  Done                    mlflow run -e build ~/mlflow-example-project
```

Re exposing mlflow (CLI and libs) methods to create Conda environments: There's a tradeoff here between flexibility and ease-of-use. I don't see parallel test execution as a compelling argument to add Conda creation as an mlflow method since you can create the conda environment in the test setup. But please let me know if there are other reasons for this functionality.

Re failed conda creation: I believe that if the conda env fails to be fully created, then it will be cleaned up.

A failed env creation:
```
(base) $: mlflow run -e build ~/mlflow-example-project 
2020/02/06 16:56:29 INFO mlflow.projects: === Creating conda environment mlflow-eb8b3286a36ea844a1c613513ade706ff2c6eeef ===
Collecting package metadata (repodata.json): done
Solving environment: failed

ResolvePackageNotFound: 
  - pandas==2.0.0

Traceback (most recent call last):
  File "/Users/avesh.singh/opt/anaconda2/bin/mlflow", line 10, in <module>
    sys.exit(cli())
  File "/Users/avesh.singh/opt/anaconda2/lib/python2.7/site-packages/click/core.py", line 764, in __call__
    return self.main(*args, **kwargs)
  File "/Users/avesh.singh/opt/anaconda2/lib/python2.7/site-packages/click/core.py", line 717, in main
    rv = self.invoke(ctx)
  File "/Users/avesh.singh/opt/anaconda2/lib/python2.7/site-packages/click/core.py", line 1137, in invoke
    return _process_result(sub_ctx.command.invoke(sub_ctx))
  File "/Users/avesh.singh/opt/anaconda2/lib/python2.7/site-packages/click/core.py", line 956, in invoke
    return ctx.invoke(self.callback, **ctx.params)
  File "/Users/avesh.singh/opt/anaconda2/lib/python2.7/site-packages/click/core.py", line 555, in invoke
    return callback(*args, **kwargs)
  File "/Users/avesh.singh/opt/anaconda2/lib/python2.7/site-packages/mlflow/cli.py", line 134, in run
    run_id=run_id
  File "/Users/avesh.singh/opt/anaconda2/lib/python2.7/site-packages/mlflow/projects/__init__.py", line 286, in run
    use_conda=use_conda, storage_dir=storage_dir, synchronous=synchronous, run_id=run_id)
  File "/Users/avesh.singh/opt/anaconda2/lib/python2.7/site-packages/mlflow/projects/__init__.py", line 162, in _run
    conda_env_name = _get_or_create_conda_env(project.conda_env_path)
  File "/Users/avesh.singh/opt/anaconda2/lib/python2.7/site-packages/mlflow/projects/__init__.py", line 451, in _get_or_create_conda_env
    conda_env_path], stream_output=True)
  File "/Users/avesh.singh/opt/anaconda2/lib/python2.7/site-packages/mlflow/utils/process.py", line 38, in exec_cmd
    raise ShellCommandException("Non-zero exitcode: %s" % (exit_code))
mlflow.utils.process.ShellCommandException: Non-zero exitcode: 1
```

We don't see this env in the list of conda environments:
```
(base) $: conda info --envs
# conda environments:
#
base                  *  /Users/avesh.singh/opt/anaconda2
mlflow-a334bf26be61bfb18b0f661cca4057c207a14948     /Users/avesh.singh/opt/anaconda2/envs/mlflow-a334bf26be61bfb18b0f661cca4057c207a14948
mlflow-dev-env           /Users/avesh.singh/opt/anaconda2/envs/mlflow-dev-env
```

**Comment by Tradunsky** [comment]
Point for exposing environment creation as mlflow command:
When we do that on CI, it usually divided to stages for user to understand better the root cause of stage failure. Currently if MLflow test stage fails, it can be either environment, mlflow or user tests. 

We want to gain more visibility for user and take control on environment re-usage, meaning, if we can create environment separately, we can reuse it during concurrent runs of mlflow as well as during retries of flaky tests or network errors.

**Comment by Tradunsky** [comment]
Regarding conda env removal on failure- that is not what I have in my case.
```mlflow run -e build ./pip-failure-model
2020/02/07 11:03:18 INFO mlflow.projects: === Creating conda environment mlflow-d3ae4f324b68093ad5e1f5e042681ee8f19cefc9 ===
Collecting package metadata (repodata.json): done
Solving environment: done

Please update conda by running

    $ conda update -n base conda


Preparing transaction: done
Verifying transaction: done
Executing transaction: done
Ran pip subprocess with arguments:
['/Users/user-name/anaconda3/envs/mlflow-d3ae4f324b68093ad5e1f5e042681ee8f19cefc9/bin/python', '-m', 'pip', 'install', '-U', '-r', '/Users/user-name/pip-failure-model/condaenv.33mg11kf.requirements.txt']
Pip subprocess output:

Pip subprocess error:
ERROR: Invalid requirement: '/Users/user-name/unknown-lib==1'
It looks like a path. File '/Users/user-name/unknown-lib==1' does not exist.


CondaEnvException: Pip failed

Traceback (most recent call last):
  File "/Users/user-name/anaconda3/bin/mlflow", line 11, in <module>
    load_entry_point('mlflow==0.9.1', 'console_scripts', 'mlflow')()
  File "/Users/user-name/anaconda3/lib/python3.7/site-packages/click/core.py", line 764, in __call__
    return self.main(*args, **kwargs)
  File "/Users/user-name/anaconda3/lib/python3.7/site-packages/click/core.py", line 717, in main
    rv = self.invoke(ctx)
  File "/Users/user-name/anaconda3/lib/python3.7/site-packages/click/core.py", line 1137, in invoke
    return _process_result(sub_ctx.command.invoke(sub_ctx))
  File "/Users/user-name/anaconda3/lib/python3.7/site-packages/click/core.py", line 956, in invoke
    return ctx.invoke(self.callback, **ctx.params)
  File "/Users/user-name/anaconda3/lib/python3.7/site-packages/click/core.py", line 555, in invoke
    return callback(*args, **kwargs)
  File "/Users/user-name/anaconda3/lib/python3.7/site-packages/mlflow/cli.py", line 139, in run
    run_id=run_id,
  File "/Users/user-name/anaconda3/lib/python3.7/site-packages/mlflow/projects/__init__.py", line 230, in run
    storage_dir=storage_dir, block=block, run_id=run_id)
  File "/Users/user-name/anaconda3/lib/python3.7/site-packages/mlflow/projects/__init__.py", line 132, in _run
    conda_env_name = _get_or_create_conda_env(project.conda_env_path)
  File "/Users/user-name/anaconda3/lib/python3.7/site-packages/mlflow/projects/__init__.py", line 462, in _get_or_create_conda_env
    conda_env_path], stream_output=True)
  File "/Users/user-name/anaconda3/lib/python3.7/site-packages/mlflow/utils/process.py", line 38, in exec_cmd
    raise ShellCommandException("Non-zero exitcode: %s" % (exit_code))
mlflow.utils.process.ShellCommandException: Non-zero exitcode: 1
```

`conda env list |grep mlflow-d3ae4f324b68093ad5e1f5e042681ee8f19cefc9`
```mlflow-d3ae4f324b68093ad5e1f5e042681ee8f19cefc9     /Users/user-name/anaconda3/envs/mlflow-d3ae4f324b68093ad5e1f5e042681ee8f19cefc9```

I know, better to check with latest version, but I had to live on the same versions that we have on production:
conda version: 4.7.12
mlflow version: 0.9.1

However, I do believe that you might have different results, if you make pip in conda to be failed to install dependency. Also, in case this is conda issue, I still do not think mlflow processes it correctly.

**Comment by AveshCSingh** [comment]
Thank you for the additional reason to expose environment creation as an MLflow command.

There's an alternate solution that allows the user to easily distinguish environment issues from MLflow or user test issues. Users can create their own Conda environments, then run their projects using `--no-conda`. I think this provides visibility and control to users who don't want to MLflow to create their Conda environment.

**Comment by Tradunsky** [comment]
Ok, thank you. Sounds as a way around for us. 

But I still would better to allow mlflow to parse its own format of MLproject rather then trying to figure out which conda environment to create.

**Comment by AveshCSingh** [comment]
I'll keep this in the back of my mind and will look for other reasons to do this / other users with this problem. If you find more reasons to decouple conda environment creation from `mlflow run` please open a new issue.

**Comment by Tradunsky** [comment]
Decoupling or just providing mlflow CLI to only create environment, would be useful when performing different mlflow CLI commands on the same model in parallel.

Thank you.
